Fix #4573 Openemr fhir token revocation

adunsulag opened the pull request:
Fixes #4573 FHIR Inferno requires you to provide an access token and a refresh token
that will be revoked in order to pass the test cases.  The developer has
to visually demonstrate revoking the token and then run the automated
test case.  This doesn't work with the trusted user mechanism we
originally had for token revokation as there isn't a way to tie the
token back to the trusted user given just the JWT.

I implemented the access token and refresh token revocation by adding a
revoked column and implementing the methods in the repositories to
handle token revocation.  I had to implement a separate isTokenRevoked
method for the access token due to the way League and OpenEMR interact
on the token revocation.  Because we store the site id inside the token
we have to allow League to validate a token and ignore the token
revocation validation check because we don't have access to the database
at that point the in execution (no site id, no database connection).

So we add the revocation check further down in the dispatch.php file in
order to handle the revocation check.

I added a lot of user interface to the Client Admin Controller to
display the list of trusted users and allow you to revoke the trusted
user right from the interface.  You can also revoke individual access
tokens. I've also added the ability to revoke individual refresh tokens 
from the admin GUI.

My next piece of work will be to add a Token Utilities button to the
client list screen.  That Token Utilities page will let you put in a
token and get back data about the token and allow you to revoke the
token from the system.  I will add that in as part of this pull request.

adunsulag commented:
Just realized I forgot to add the SQL for the initial database stuff.  Since this is a work in progress PR and more to get eyeballs on this I will work on that tomorrow.

bradymiller commented on sql/6_0_0-to-6_1_0_upgrade.sql:
```diff
@@ -932,3 +932,21 @@ INSERT INTO list_options (list_id,option_id,title,seq,is_default,activity, notes
+    `user_id` VARCHAR(40),
+    `client_id` VARCHAR(80),
+    `token` VARCHAR(128) NOT NULL,
+    `expiry` DATETIME,
```
default date stuff to NULL

bradymiller replied:
not addressed yet (they were addressed in database.sql though)

bradymiller commented on sql/6_0_0-to-6_1_0_upgrade.sql:
```diff
@@ -932,3 +932,21 @@ INSERT INTO list_options (list_id,option_id,title,seq,is_default,activity, notes
+    `id` BIGINT(20) NOT NULL AUTO_INCREMENT,
+    `user_id` VARCHAR(40),
+    `client_id` VARCHAR(80),
+    `token` VARCHAR(128) NOT NULL,
```
should the above 2 fields be matching api_token in the default settings (ie. null)?

adunsulag replied:
I'll make them match though I don't understand why they defaulted to NULL in the first place as things like expiry should always be filled out in my view.

bradymiller replied:
not addressed yet (they were addressed in database.sql though)

bradymiller commented on src/Common/Auth/OpenIDConnect/Repositories/AccessTokenRepository.php:
```diff
@@ -71,4 +100,30 @@ public function getNewToken(ClientEntityInterface $clientEntity, array $scopes,
+
+    public function getTokenByToken($token)
+    {
+        $sql = "SELECT * FROM api_token WHERE id = ? ";
```
is this query correct or is it supposed to be WHERE `token` = ...?

adunsulag replied:
good catch, wonder why it was working haha.

bradymiller commented on sql/6_0_0-to-6_1_0_upgrade.sql:
```diff
@@ -932,3 +932,21 @@ INSERT INTO list_options (list_id,option_id,title,seq,is_default,activity, notes
+    `expiry` DATETIME,
+    `revoked` TINYINT(1) NOT NULL DEFAULT 0 COMMENT '1=revoked,0=not revoked',
+    PRIMARY KEY (`id`),
+    INDEX `api_refresh_token_token_idx` (`token`),
```
should token be unique like in api_token table?

bradymiller commented on src/FHIR/SMART/ClientAdminController.php:
```diff
@@ -130,12 +147,68 @@ public function editAction($clientId, $request)
+        foreach ($trustedUsers as $user) {
+            // TODO: do we need to optimize this query?
+            if (UuidRegistry::isValidStringUUID($user['user_id'])) {
+                $registryRecord = UuidRegistry::getRegistryRecordForUuid($user['user_id']);
```
![2DV](https://user-images.githubusercontent.com/278968/128997318-9a760205-8d47-49d6-8453-7442da7848f9.gif)

bradymiller replied:
This is a neat use of the uuid registry table. But couldn't you just search directly in the patient_data table uuid column?

adunsulag replied:
@bradymiller Problem is that the user_id can  beither a patient_data.uuid or a users.uuid so I would have to search both tables first.  That's why I hit the registry to figure out what kind of uuid it is.

adunsulag commented:
Here are screenshots for this pull request:  First off more Inferno checkmarks!
![image](https://user-images.githubusercontent.com/228117/129091254-95576192-05d7-4564-bba9-066f55dd61d3.png)

There is now a Token Tools button in the top right corner of the client apps page.
![image](https://user-images.githubusercontent.com/228117/129091413-fb2787e7-ddb4-4b6a-b7ae-57d007c1da35.png)

Clicking on that opens up the token tools page where you can paste in a refresh or access token to get details on the token and revoke the token if needed:
![image](https://user-images.githubusercontent.com/228117/129091809-7938bd45-b062-4454-be45-7d9bf0a864ce.png)
![image](https://user-images.githubusercontent.com/228117/129091651-6526cb57-4821-431d-bcf4-9167e34f5359.png)

Hitting the revoke token will prevent the token from being used by api clients:
![image](https://user-images.githubusercontent.com/228117/129091963-c7d409c7-de61-4b89-ab71-d2615d13f390.png)

I also added to each api client a display of all of the authorized users and access/refresh tokens for a api client.  You can revoke individual tokens there as well as revoke trusted users from the interface:
![image](https://user-images.githubusercontent.com/228117/129092311-a3ec6035-6d91-4015-831f-cc6989ff038d.png)
![image](https://user-images.githubusercontent.com/228117/129092352-1651a371-a9ca-4898-b7ec-ec70a71c09a3.png)

I also added a few more fields such as registration date, json web keys, uris, etc.
![image](https://user-images.githubusercontent.com/228117/129092515-40a59cf0-5b06-4c95-a1b7-dfec5841c406.png)

bradymiller commented on src/Common/Auth/OpenIDConnect/Repositories/RefreshTokenRepository.php:
```diff
@@ -13,28 +13,83 @@
+        return $result ?? null; // if the result set is not empty then its been revoked, otherwise its a good token
+    }
+
+    public function getTokenByToken($token)
```
this function is the same as the above getRefreshTokenByToken() function

adunsulag commented:
Hey @bradymiller I believe I've fixed the issues that you identified.  If it looks good to you, I'll bring it in.  Thanks for the second set of eyes!

bradymiller commented on src/FHIR/SMART/ClientAdminController.php:
```diff
@@ -558,4 +843,273 @@ private function renderFooter()
+        <li><?php echo xlt("Token DB Id"); ?>: <?php echo $databaseRecord['id']; ?></li>
+        <li><?php echo xlt("Token Status"); ?>: <?php echo $parts['status']; ?></li>
+        <li><?php echo xlt("Token Type"); ?>: <?php echo $parts['token_type']; ?></li>
+        <li><?php echo xlt("Token Expiration"); ?>: <?php echo $databaseRecord['expiry']; ?></li>
```
above 5 lines need a text()

bradymiller commented on src/FHIR/SMART/ClientAdminController.php:
```diff
@@ -558,4 +843,273 @@ private function renderFooter()
+        <li><?php echo xlt("Token Expiration"); ?>: <?php echo $databaseRecord['expiry']; ?></li>
+        <li><?php echo xlt("User UUID"); ?>:
+        <a href="<?php echo attr($parts['user_link']); ?>">
+            <?php echo $databaseRecord['user_id']; ?>
```
text() in above line

bradymiller commented on src/FHIR/SMART/ClientAdminController.php:
```diff
@@ -558,4 +843,273 @@ private function renderFooter()
+        </a>
+        </li>
+        </ul>
+        <pre><?php echo json_encode($parts, JSON_PRETTY_PRINT); ?></pre>
```
should wrap above with a text()
(at least I think we should since this is ouput in html context)

bradymiller commented on src/RestControllers/AuthorizationController.php:
```diff
@@ -140,97 +142,17 @@ private function configKeyPairs(): void
+            $serverException = OAuthServerException::serverError("Security error - problem with authorization server keys.", $exception);
             SessionUtil::oauthSessionCookieDestroy();
-            $this->emitResponse($exception->generateHttpResponse($response));
-            exit;
```
very cool to see this stuff get refactored into a separate class. Just making sure it will still force an exit from script after catching the exception and emitting the response (noted the `exit;` was removed)

bradymiller commented on src/RestControllers/AuthorizationController.php:
```diff
@@ -655,6 +577,7 @@ public function getAuthorizationServer(): AuthorizationServer
             $responseType->markIsAuthorizationGrant(); // we have specific SMART responses for an authorization grant.
         }
 
+        $this->logger->debug("key log ", ['privateKey' => $this->privateKey, 'passphrase' => $this->passphrase, 'encryptionKey' => $this->oaEncryptionKey]);
```
I prob wouldn't log these actual values in a log file (since would consider them very sensitive info from security perspective). Since a novice user could have debug on at some point and potentially pass their log somewhere in unsecure fashion.

bradymiller replied:
maybe just show if they are empty or not if helpful for debug purposes

adunsulag replied:
Wow... that was only for my own debugging purposes.  Can't believe I left that in there.  I'm really, really glad you caught that.

bradymiller commented on src/RestControllers/AuthorizationController.php:
```diff
@@ -1190,55 +1113,42 @@ public function tokenIntrospection(): void
+//                    $token_hint = 'access_token';
+//                } else {
+//                    $token_hint = 'refresh_token';
+//                }
```
would remove rather than comment out

bradymiller commented on src/Common/Auth/OAuth2KeyConfig.php:
```diff
@@ -0,0 +1,161 @@
+use OpenEMR\Common\Crypto\CryptoGen;
+use OpenEMR\Common\Utils\RandomGenUtils;
+
+class OAuth2KeyConfig
```
![2DV-2](https://user-images.githubusercontent.com/278968/129138679-181f9f81-f0fc-4f0f-8a46-57f7b1e953c0.gif)

bradymiller commented:
hi @adunsulag , noted a couple more minor issues. I promise that is it. After issues are addressed, then bring it in :)
